### ISSUE TITLE
Replace _Py_IDENTIFIER() with statically initialized objects.

### ISSUE THREAD

**Issue by ericsnowcurrently** [python/cpython#90699]
BPO | [46541](https://bugs.python.org/issue46541)
--- | :---
Nosy | @vstinner, @tiran, @ericsnowcurrently, @serhiy-storchaka, @corona10, @erlend-aasland, @kumaraditya303, @notarealdeveloper
PRs | <li>python/cpython#30928</li><li>python/cpython#30941</li><li>python/cpython#31261</li><li>python/cpython#31344</li><li>python/cpython#31346</li><li>python/cpython#31351</li><li>python/cpython#31358</li><li>python/cpython#31363</li><li>python/cpython#31364</li><li>python/cpython#31372</li><li>python/cpython#31375</li><li>python/cpython#31376</li><li>python/cpython#30310</li><li>python/cpython#31468</li><li>python/cpython#31475</li><li>python/cpython#31599</li><li>python/cpython#31608</li><li>python/cpython#31609</li><li>python/cpython#31683</li><li>python/cpython#32063</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/ericsnowcurrently'
closed_at = None
created_at = <Date 2022-01-26.22:01:04.421>
labels = ['interpreter-core', '3.11']
title = 'Replace _Py_IDENTIFIER() with statically initialized objects.'
updated_at = <Date 2022-03-23.15:52:58.371>
user = 'https://github.com/ericsnowcurrently'
```

bugs.python.org fields:
```python
activity = <Date 2022-03-23.15:52:58.371>
actor = 'eric.snow'
assignee = 'eric.snow'
closed = False
closed_date = None
closer = None
components = ['Interpreter Core']
creation = <Date 2022-01-26.22:01:04.421>
creator = 'eric.snow'
dependencies = []
files = []
hgrepos = []
issue_num = 46541
keywords = ['patch']
message_count = 28.0
messages = ['411799', '411800', '411910', '411987', '412229', '412405', '412488', '412868', '413047', '413086', '413246', '413274', '413311', '413312', '413323', '413333', '413334', '413337', '413355', '413374', '413384', '413653', '414178', '414182', '414265', '414274', '414534', '415881']
nosy_count = 8.0
nosy_names = ['vstinner', 'christian.heimes', 'eric.snow', 'serhiy.storchaka', 'corona10', 'erlendaasland', 'kumaraditya', 'notarealdeveloper']
pr_nums = ['30928', '30941', '31261', '31344', '31346', '31351', '31358', '31363', '31364', '31372', '31375', '31376', '30310', '31468', '31475', '31599', '31608', '31609', '31683', '32063']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = None
url = 'https://bugs.python.org/issue46541'
versions = ['Python 3.11']
```

</p></details>

* PR: gh-98956

* PR: gh-98957

* PR: gh-99010

* PR: gh-99012

* PR: gh-99054

* PR: gh-99067

* PR: gh-99210

* PR: gh-99236

* PR: gh-101185

**Comment by kumaraditya303** [comment]
With https://github.com/python/cpython/pull/93597, the global identifier strings are interned and now can be used in-place of other cached interned strings to avoid allocations and fast comparisons with pointers.

**Comment by kumaraditya303** [comment]
I am working on some of these so you can leave it to me to sort out.

**Comment by kumaraditya303** [comment]
Out of these modules, `asyncio` and `json` already use `internal` C API for best performance so I am thinking of using `_Py_ID()` for these modules. I agree that we should avoid using `internal` C API in extensions but these modules are special as they implemented in C for best performance so using `_Py_ID()` seems better.

**Comment by kumaraditya303** [comment]
ctypes and datetime are similar cases which already use internal C API so I'll change it to use `_Py_ID`.

**Comment by ericsnowcurrently** [comment]
I'm okay with that, though ideally we would use public API as much as possible and limit `_Py_ID()` to performance-sensitive code.  That said, I favor getting this done over being excessively careful about how much we use `_Py_ID()`. 🙂

**Comment by kumaraditya303** [comment]
> I'm okay with that, though ideally we would use public API as much as possible and limit _Py_ID() to performance-sensitive code. That said, I favor getting this done over being excessively careful about how much we use _Py_ID(). 🙂

practicality beats purity :) We can of course reconsider this when these modules will be ported to multiphase heap types.

**Comment by kumaraditya303** [comment]
@ericsnowcurrently: I will finish this today itself, I have the code ready for next final PR, waiting for https://github.com/python/cpython/pull/99054 to be merged to avoid conflict.

**Comment by kumaraditya303** [comment]
FYI, once https://github.com/python/cpython/pull/99067 is merged, I'll disable using `_Py_IDENTIFIER` in core code. With that completed I consider this done and deprecation/removal for `_Py_IDENTIFIER` can be discussed separately.

**Comment by ericsnowcurrently** [comment]
We will address Programs/_testembed.c in the issue/PR where we remove `_Py_IDENTIFIER()` (or decide to not).

**Comment by ericsnowcurrently** [comment]
Thanks for doing all that work, @kumaraditya303!

**Comment by ericsnowcurrently** [comment]
We can close this once gh-99210 lands.

**Comment by ericsnowcurrently** [comment]
This side of things is done.  We can create a new issue about maybe getting rid of `_Py_IDENTIFIER()` and what to do about third-party usage in the community.

**Comment by kumaraditya303** [comment]
Thanks to everyone!

@ericsnowcurrently: It would be nice to update to c analyzer as a lot of globals are removed as part of this.
